### PR TITLE
Add Callback logic for Singleton

### PR DESCRIPTION
**Pull request by eluchsinger**
PR of #1213

### CODE REVIEW

**Comment by DDReaper** [comment]
Potential conflicts with #1208

**Comment by StephenHodgson** [comment]
I'm not a fan of this change.  @eluchsinger have you read back through the other Singleton discussions?

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 1]
```diff
@@ -1,53 +1,27 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: Missing license header

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 2]
```diff
@@ -1,53 +1,27 @@
...
-// Licensed under the MIT License. See LICENSE in the project root for license information.
-
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: We probably don't need all of these

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 3]
```diff
@@ -1,53 +1,27 @@
...
 using UnityEngine;
 
 namespace HoloToolkit.Unity
 {
     /// <summary>
     /// Singleton behaviour class, used for components that should only have one instance
     /// </summary>
-    /// <typeparam name="T">The Singleton Type</typeparam>
+    /// <typeparam name="T"></typeparam>
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: Why remove the xml docs?

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 4]
```diff
@@ -1,53 +1,27 @@
...
         private static T instance;
+        private static List<Action> actionsWaitingForInitialization = new List<Action>();
 
-        /// <summary>
-        /// Returns the Singleton instance of the classes type.
-        /// If no instance is found, then we search for an instance
-        /// in the scene.
-        /// If more than one instance is found, we throw an error and
-        /// no instance is returned.
-        /// </summary>
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: Why remove the xml summary?

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 5]
```diff
@@ -1,53 +1,27 @@
...
-                    if (objects.Length == 1)
-                    {
-                        instance = objects[0];
-                    }
-                    else if (objects.Length > 1)
-                    {
-                        Debug.LogErrorFormat("Expected exactly 1 {0} but found {1}.", typeof(T).ToString(), objects.Length);
-                    }
-                }
                 return instance;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: How do you handle the case if the instance is null?

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 6]
```diff
@@ -1,53 +1,27 @@
...
             }
         }
 
-        private static bool searchForInstance = true;
-
-        public static void AssertIsInitialized()
-        {
-            Debug.Assert(IsInitialized, string.Format("The {0} singleton has not been initialized.", typeof(T).Name));
-        }
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: Why remove the assert?

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 7]
```diff
@@ -60,45 +34,46 @@ public static bool IsInitialized
...
+                actionsWaitingForInitialization = null;
             }
         }
 
-        /// <summary>
-        /// Base OnDestroy method that destroys the Singleton's unique instance.
-        /// Called by Unity when destroying a MonoBehaviour. Scripts that extend
-        /// Singleton should be sure to call base.OnDestroy() to ensure the
-        /// underlying static Instance reference is properly cleaned up.
-        /// </summary>
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: Why remove the xml summary?

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 8]
```diff
@@ -60,45 +34,46 @@ public static bool IsInitialized
...
+        /// <param name="action">Action to perform when the singleton is initialized.</param>
+        public static void WhenInitialized(Action action)
+        {
+            if (IsInitialized)
+            {
+                action();
+            }
+            else
+            {
+                actionsWaitingForInitialization.Add(action);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: How is this used in the rest of the Singleton instances.  Could you explain the _flow_ of events?

File: Assets/HoloToolkit/Utilities/Scripts/Singleton.cs [hunk 9]
```diff
@@ -1,6 +1,8 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Singleton.cs]: Not sure why it seems to think there's a change to this line still.

**Pull request closed without merging**